### DISCUSSION TITLE
Error [TypeError]: Cannot read properties of undefined (reading 'forEach')

### DISCUSSION THREAD

**Discussion by zhakezhanovv** [vercel/next.js#50520] [App Router]
The error occurs when starting a server during a migration from pages to app folder

<img width="1057" alt="Screenshot 2023-05-30 at 3 38 12 PM" src="https://github.com/vercel/next.js/assets/97029797/2cf6a30b-231e-4d7d-8cf7-7690fa196caf">

**Comment by icyJoseph** [comment]
Uff unless other people are also experiencing this, this probably needs more information.

- I guess you don't actually have `forEach` calls, in your codebase?
- What is app/page.js using in terms of API? is it using Pages directory only APIs?

Any other info you could share would be good.

**Reply by zhakezhanovv** [reply]
Thanks for the quick reply! 

1) Correct, there are no `forEach` calls in my codebase.
2) For now, app/page.js is not using any APIs. It is just a dummy page. 
`export default function Page() {
  return <div>Page</div>;
}
`.

**Reply by icyJoseph** [reply]
Alright, that's good. What about `app/layout.tsx`? Also, you made sure to remove `pages/index.js` right? Those would otherwise crash with each other.

**Reply by zhakezhanovv** [reply]
Layout.js looks like as follows: 

`export const metadata = {
  title: 'Next.js',
  description: 'Generated by Next.js',
};

export default function RootLayout({ children }) {
  return (
    <html lang='en'>
      <body>{children}</body>
    </html>
  );
}`

Yes, index.js was removed. Basically, I started with simple things for testing purposes.

**Reply by icyJoseph** [reply]
Right, but I think you have to add a html/body tags in the layout, it is expected:

- https://nextjs.org/docs/app/building-your-application/routing/pages-and-layouts#root-layout-required

> Good to know:
> 
> - The app directory must include a root layout.
> - The root layout must define `<html>` and `<body>` tags since Next.js does not automatically create them.
> - You can use the [built-in SEO support](https://nextjs.org/docs/app/building-your-application/optimizing/metadata) to manage <head> HTML elements, for example, the <title> element.
> - You can use [route groups](https://nextjs.org/docs/app/building-your-application/routing/route-groups) to create multiple root layouts. See an [example here](https://nextjs.org/docs/app/building-your-application/routing/route-groups#creating-multiple-root-layouts).
> - The root layout is a [Server Component](https://nextjs.org/docs/getting-started/react-essentials) by default and can not be set to a [Client Component](https://nextjs.org/docs/getting-started/react-essentials#client-components).


Could you double check if adding those fixes the problem?

**Reply by zhakezhanovv** [reply]
Oh, for some reason the code that I pasted in the previous reply did not include tags. But, yes, they were included in the code.

<img width="1057" alt="Screenshot 2023-05-30 at 4 15 51 PM" src="https://github.com/vercel/next.js/assets/97029797/17f603f1-1e15-40b0-aa67-cbaf9c899485">

**Reply by icyJoseph** [reply]
I think Github blanks those... I had just tried, and didn't see the same issue.

So all you have is a RootLayout and a simple one tag page.

What Node.js version have you got?

**Reply by zhakezhanovv** [reply]
Yes, only layout.js and page.js.

I am using v20.2.0.

**Reply by icyJoseph** [reply]
Mmm odd I cannot reproduce this. I was hoping your node version had some mismatching API, or missing API rather.

If you do a `yarn create next-app delete-this-project` and go in, it runs without issue right? (Or with `npm create-next-app`)

**Reply by zhakezhanovv** [reply]
Yes, for a newly created application, it works fine. So I am wondering if this is an issue with an old app or next.js bug.

**Reply by icyJoseph** [reply]
Without looking at more code, or the project structure and dependencies, this looks kind of difficult to debug.
